src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,8 +153,218 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(other.dataset),
+    delta(other.delta),
+    parent(NULL),
+    left(NULL),
+    right(NULL),
+    indices(other.indices),
+    l2NormsSquared(other.l2NormsSquared),
+    centroid(other.centroid),
+    basisVector(other.basisVector),
+    splitPointIndex(other.SplitPointIndex()),
+    numColumns(other.NumColumns()),
+    l2Error(other.L2Error()),
+    frobNormSquared(other.FrobNormSquared()),
+    localDataset(other.localDataset && other.parent == NULL)
+{
+  // Performing a deep copy of the dataset.
+  arma::mat cpy = other.GetDataset();
+  dataset = &cpy;
```

Comment:
This is actually even more unsafe---`cpy` is a local variable and thus will go out of scope at the time this constructor ends, leaving `dataset` a pointer to deallocated memory.  I think maybe it would be helpful to spend some time reading about heap and stack allocations and memory management in C++---if you search for some nice resource, I am sure there are places that can explain it better than I can.

In our case, what we want to do is create a new stack-allocated matrix that has copied all the contents of `other.dataset` to use as `dataset`, so, in the initialization list, we can use `dataset(new arma::mat(*other.dataset))`, and then we have to be sure that `localDataset` is set correctly.

Reply:
Got it, will do. The reason I took that approach was because @walragatver specified that in one his earlier comments 
 (https://github.com/mlpack/mlpack/pull/2023#discussion_r331147450)